llvm/labs/lab2/lebedeva_ksenia/LebedevaBitwiseShift.cpp

```diff
@@ -0,0 +1,74 @@
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+using namespace llvm;
+
+class BitwiseShift : public PassInfoMixin<BitwiseShift> {
+public:
+  PreservedAnalyses run(llvm::Function &func,
+                        llvm::FunctionAnalysisManager &aManager) {
+    bool changed = false;
+    for (auto &basicBlock : func) {
+      for (auto InstIt = basicBlock.begin(); InstIt != basicBlock.end();) {
+        Instruction &instr = *InstIt++;
+        if (instr.getOpcode() != Instruction::Mul) {
+          continue;
+        }
+        auto *Op = dyn_cast<BinaryOperator>(&instr);
+        if (!Op) {
+          continue;
+        }
+
+        Value *leftVal = Op->getOperand(0);
+        Value *rightVal = Op->getOperand(1);
+        int leftLog = getLog2(leftVal);
+        int rightLog = getLog2(rightVal);
+
+        if (rightLog < leftLog) {
+          std::swap(leftLog, rightLog);
+          std::swap(leftVal, rightVal);
+        }
+        if (rightLog >= 0) {
+          IRBuilder<> Builder(Op);
+          Value *NewVal = Builder.CreateShl(
+              leftVal, ConstantInt::get(Op->getType(), rightLog));
+          instr.replaceAllUsesWith(NewVal);
+          InstIt = instr.eraseFromParent();
+          changed = true;
+        }
+      }
+    }
+    return changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+  }
+
+private:
+  int getLog2(llvm::Value *Op) {
+    if (auto *CI = dyn_cast<ConstantInt>(Op)) {
+      return CI->getValue().exactLogBase2();
+    }
+    return -1;
```

Comment:
From mathematical point of view, `log2(1/2) == 1`, so `-1` is not good value to be used as an indicator of the error. I'd suggest switching for std::optional here (https://en.cppreference.com/w/cpp/utility/optional). It helps to indicate empty return value